source/_posts/2020-02-18-wasm-ewasm-what-and-why.md

```diff
@@ -0,0 +1,120 @@
+
+![wasm editor](/assets/images/wasm_explorer_online_app.png)
+
+From what I've read, I believe one of the biggest ideas behind Wasm; is to make it possible to run media-rich game engines, and support such graphics-heavy games in-browser, without the use of plug-ins. It also has non-web applications such as the Internet of Things, mobile apps and JavaScript virtual machines.
+
+<iframe width="680" height="382" src="https://www.youtube.com/embed/MaJCfdmr9Wg" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+
+
+If you want to get started with Wasm, you can download a precompiled toolchain to compile C/C++ to WebAssembly by running the following:
+
+```bash
+$ git clone https://github.com/emscripten-core/emsdk.git
+$ cd emsdk
+$ ./emsdk install latest
+$ ./emsdk activate latest
+```
+
+
+## Wasm to eWasm (Ethereum flavoured WebAssembly)
+
+Great! &nbsp; Now that we have Wasm introduced, explained, and out-of-the-way, we can move onto the (arguably) more important topic – ***eWasm***!  So.. what exactly is eWasm?
+
+Simply put; eWasm is a restricted subset of Wasm to be used for contracts in Ethereum.  Much like Wasm, one of the biggest goals of eWasm is to be fast & efficient.  To truly distinguish Ethereum as the *"World Computer",* we need to have a super performant VM. The current architecture of the VM is one of the greatest blockers to raw performance.  
+
+As I mentioned in the Wasm section above; WebAssembly aims to execute at near native speed by taking advantage of common hardware capabilities available on a wide range of platforms. This will open the door for Ethereum to a wide array of uses that require performance/throughput.
+
+Security is another key goal.  With the added performance gains from eWasm we will be able to implement parts of Ethereum such as the precompiled contract in the VM itself which will minimise our trusted computing base.  WebAssembly is currently being designed as an open standard by a W3C Community Group and is actively being developed by engineers from Mozilla, Google, Microsoft, and Apple
+
+
+
+## eWasm Goals
+
+Goals of the eWasm project:
+
+ - To provide a specification of eWasm contract semantics and the Ethereum interface.
+ - To provide an EVM transcompiler, preferably as an eWasm contract.
+ - To provide a metering injector*, preferably as an eWasm contract.
+ - To provide a VM implementation for executing eWasm contracts.
+ - To implement an eWasm backend in the Solidity compiler.
+ - To provide a library and instructions for writing contracts in Rust.
+ - To provide a library and instructions for writing contracts in C.
+
+*Metering injector is a transformation tool inserting metering code to an eWasm contract.
```

Comment:
This one seems a little out of place as it hasn't been mentioned before and there's no additional info on why this is mentioned here.